File: crates/normalize-languages/tests/fixtures/ruby/sample.rb
require 'json'
require 'set'

# A simple stack data structure
class Stack
  def initialize
    @data = []
  end

  def push(item)
    @data.push(item)
    self
  end

  def pop
    if @data.empty?
      raise "Stack is empty"
    end
    @data.pop
  end

  def peek
    @data.last
  end

  def empty?
    @data.empty?
  end

  def size
    @data.size
  end
end

# Classify a number as negative, zero, or positive
def classify(n)
  if n < 0
    :negative
  elsif n == 0
    :zero
  else
    :positive
  end
end

# Sum elements in an array that satisfy a predicate
def sum_if(arr, &block)
  total = 0
  arr.each do |x|
    total += x if block.call(x)
  end
  total
end

stack = Stack.new
stack.push(1).push(2).push(3)
puts stack.pop
puts classify(-5)
puts sum_if([1, 2, 3, 4, 5]) { |x| x.even? }

File: crates/normalize-languages/tests/fixtures/elixir/sample.ex
defmodule MathUtils do
  alias Enum, as: E

  @doc "Classify a number as :negative, :zero, or :positive"
  def classify(n) do
    cond do
      n < 0 -> :negative
      n == 0 -> :zero
      true -> :positive
    end
  end

  @doc "Sum elements matching the predicate"
  def sum_if(list, predicate) do
    Enum.reduce(list, 0, fn x, acc ->
      if predicate.(x), do: acc + x, else: acc
    end)
  end

  def sum_evens(numbers) do
    sum_if(numbers, fn n -> rem(n, 2) == 0 end)
  end
end

defmodule Stack do
  import Enum, only: [reverse: 1]

  defstruct items: []

  def new(), do: %Stack{}

  def push(%Stack{items: items}, item) do
    %Stack{items: [item | items]}
  end

  def pop(%Stack{items: []}) do
    {:error, :empty}
  end

  def pop(%Stack{items: [head | tail]}) do
    {:ok, head, %Stack{items: tail}}
  end

  def peek(%Stack{items: []}), do: nil
  def peek(%Stack{items: [head | _]}), do: head

  def to_list(%Stack{items: items}), do: reverse(items)
end
